runtime/gc_stats.c

```diff
@@ -0,0 +1,118 @@
+/*                                                                        */
+/*             Gabriel Scherer, projet Partout, INRIA Saclay              */
+/*                                                                        */
+/*   Copyright 2022 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+#include "caml/shared_heap.h"
+#include "caml/gc_stats.h"
+
+Caml_inline intnat intnat_max(intnat a, intnat b) {
+  return (a > b ? a : b);
+}
+
+void caml_accum_heap_stats(struct heap_stats* acc, const struct heap_stats* h)
+{
+  acc->pool_words += h->pool_words;
+  acc->pool_max_words = intnat_max(acc->pool_max_words, acc->pool_words);
+  acc->pool_max_words = intnat_max(acc->pool_max_words, h->pool_max_words);
+  acc->pool_live_words += h->pool_live_words;
+  acc->pool_live_blocks += h->pool_live_blocks;
+  acc->pool_frag_words += h->pool_frag_words;
+  acc->large_words += h->large_words;
+  acc->large_max_words = intnat_max(acc->large_max_words, acc->large_words);
+  acc->large_max_words = intnat_max(acc->large_max_words, h->large_max_words);
+  acc->large_blocks += h->large_blocks;
+}
+
+void caml_remove_heap_stats(struct heap_stats* acc, const struct heap_stats* h)
+{
+  acc->pool_words -= h->pool_words;
+  acc->pool_live_words -= h->pool_live_words;
+  acc->pool_live_blocks -= h->pool_live_blocks;
+  acc->pool_frag_words -= h->pool_frag_words;
+  acc->large_words -= h->large_words;
+  acc->large_blocks -= h->large_blocks;
+}
+
+void caml_accum_alloc_stats(struct alloc_stats* acc, const struct alloc_stats* s) {
+  acc->minor_words += s->minor_words;
+  acc->promoted_words += s->promoted_words;
+  acc->major_words += s->major_words;
+  acc->minor_collections += s->minor_collections;
+  acc->forced_major_collections += s->forced_major_collections;
+}
+
+void caml_collect_alloc_stats_sample(
+  caml_domain_state *local,
+  struct alloc_stats *sample)
+{
+  sample->minor_words = local->stat_minor_words;
+  sample->promoted_words = local->stat_promoted_words;
+  sample->major_words = local->stat_major_words;
+  sample->minor_collections = local->stat_minor_collections;
+  sample->forced_major_collections = local->stat_forced_major_collections;
+}
+
+/* The "sampled stats" of a domain are a recent copy of its
+   domain-local stats, accessed without synchronization and only
+   updated ("sampled") during stop-the-world events -- each minor
+   collection. */
```

Comment:
Ditto as for the header file: we also collect the samples at domain termination.